homeassistant/config.py

```diff
@@ -1030,7 +1030,11 @@ async def merge_packages_config(
     ] = _log_pkg_error,
 ) -> dict:
     """Merge packages into the top-level configuration. Mutate config."""
-    PACKAGES_CONFIG_SCHEMA(packages)
+    try:
+        PACKAGES_CONFIG_SCHEMA(packages)
+    except vol.Error as err:
+        raise HomeAssistantError(f"Error validating package config: {err}") from err
```

Comment:
I think this function is expected to use `_log_pkg_error` to record package config errors. See example on line 1045.

Since this validation validates all of the packages it may be hard to determine which package(s) is invalid. If we can't do that, I think the caller of `merge_packages_config` should handle `vol.Invalid` and record an error for the whole packages item as the caller does for other config items.

Reply:
OK... My rationale was to keep the current behavior (which would simply bail out of the whole config loading) and make it compatible to the currently existing activation of recovery mode, which is supposedly intended as that is what happens for other severe errors on config parse/load. The bootstrap code fetches and treats `HomeAssistantError` and activates recovery mode; on other exceptions it just fails and aborts (see https://github.com/home-assistant/core/blob/6bc36666b1551c9fdc567b02d122398fd687e787/homeassistant/bootstrap.py#L148).

I did explicitly _not_ want to alter the startup behavior. Clearly there are alternatives; for instance we could simply skip the whole "packages" section with a nice friendly error message in the logs (and not activate recovery mode at all), or even follow @emontnemery's suggestion and remove the global schema check of the full packages config and move it inside the for loops to only skip erroneous packages and continue with the others. I am open for both alternatives, I'd just like to know what would be more acceptable, and which checks there are in place besides key_needs_to_be_a_slug and value_needs_to_be_a_dict (if any). I was unable to cause other validation errors, but you can't be too sure ;)

Reply:
Bootstrap isn't the only place where `merge_packages_config` is indirectly used. It's also used in the check config helper. We need to keep the intended behavior of all users of `merge_packages_config`.

Reply:
I think the `PACKAGES_CONFIG_SCHEMA` is there to make subsequent validation easier by validating that the main structure of the packages config has the correct types and format.

It will be hard to validate the parts if some baseline isn't established, eg that the container is a dict. If the config container is completely incorrect we need to abort early anyway.

Reply:
> Bootstrap isn't the only place where `merge_packages_config` is indirectly used. It's also used in the check config helper. We need to keep the intended behavior of all users of `merge_packages_config`.

I know. Without patch:
![grafik](https://github.com/home-assistant/core/assets/57918757/f7290118-7ddf-48aa-bb3f-728268acbb06)

With patch:
![grafik](https://github.com/home-assistant/core/assets/57918757/f29423a7-07ad-40a2-be93-19d1fd069a1d)

Works for me 😉 

Startup log without patch (I skipped traceback elements to keep it short; you get the idea):
```
  File "/workspaces/homeassistant-core/homeassistant/runner.py", line 154, in setup_and_run_hass
    hass = await bootstrap.async_setup_hass(runtime_config)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/homeassistant-core/homeassistant/bootstrap.py", line 147, in async_setup_hass
    config_dict = await conf_util.async_hass_config_yaml(hass)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/homeassistant-core/homeassistant/config.py", line 505, in async_hass_config_yaml
    await merge_packages_config(hass, config, core_config.get(CONF_PACKAGES, {}))
  File "/workspaces/homeassistant-core/homeassistant/config.py", line 1033, in merge_packages_config
    PACKAGES_CONFIG_SCHEMA(packages)
  File "/workspaces/homeassistant-core/homeassistant/helpers/config_validation.py", line 588, in verify
    return cast(dict, schema(value))
                      ^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/voluptuous/schema_builder.py", line 272, in __call__
    return self._compiled([], data)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/voluptuous/schema_builder.py", line 595, in validate_dict
    return base_validate(path, iteritems(data), out)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/voluptuous/schema_builder.py", line 433, in validate_mapping
    raise er.MultipleInvalid(errors)
voluptuous.error.MultipleInvalid: expected a dictionary for dictionary value @ data['horst']
```


Startup log with patch:
```
2024-01-17 06:01:03.404 ERROR (MainThread) [homeassistant.bootstrap] Failed to parse configuration.yaml: Error validating package config: expected a dictionary for dictionary value @ data['horst']. Activating recovery mode
```


> It will be hard to validate the parts if some baseline isn't established, eg that the container is a dict. If the config container is completely incorrect we need to abort early anyway.

That depends on what is actually verified. I know of three properties:
1. `packages` is a `dict`
2. Eeach key of `packages` is a slug
3. Each value of `packages` is a dict

To check these properties kind of more explicitely does not seem too complicated.

Reply:
> My rationale was to keep the current behavior (which would simply bail out of the whole config loading) and make it compatible to the currently existing activation of recovery mode, which is supposedly intended as that is what happens for other severe errors on config parse/load. 

We only active recovery mode automatically for severe issues with the configuration, for example failure to setup frontend.

A problem with package configuration which does not otherwise block startup does not warrant entering recovery mode.

Hence my request to start with improving the error message if `PACKAGES_CONFIG_SCHEMA` raises, and if that's difficult, modify `PACKAGES_CONFIG_SCHEMA`.

Did you look into the functions I mentioned which help format `vol.Invalid` to make it user readable?

Reply:
> We only active recovery mode automatically for severe issues with the configuration, for example failure to setup frontend.
> 
> A problem with package configuration which does not otherwise block startup does not warrant entering recovery mode.

I guess that makes sense; however, currently the whole application doesn't start at all in these cases. So having it run in recovery mode is at least an improvement 😉 


> Hence my request to start with improving the error message if `PACKAGES_CONFIG_SCHEMA` raises, and if that's difficult, modify `PACKAGES_CONFIG_SCHEMA`.

I started messing with the validation a little more. One issue is that the control flow between bootstrap and config check is quite different (something I don't plan to address in this MR, though...): while bootstrapping package merge happens before an additional validation against the full core config schema, during config check it's the other way around. There the full schema validation happens first and afterwards package merge is performed.

That said, my current local WIP version looks like this:
1. Simplified `PACKAGES_CONFIG_SCHEMA` that bails out during package merge only on super severe validation errors (which would then trigger recovery mode; yet I was unable to provoke this; don't exactly know yet what I can do to my config files in order to actually trigger _that_) (that also affects `CORE_CONFIG_SCHEMA` btw, but as this is always used together with merge in some way, I think it's fine)
2. Check each key in `packages` for being a slug and each value being a dict and produce nice sweet error messages in case something goes wrong; something like `2024-01-17 15:17:21.716 ERROR (MainThread) [homeassistant.config] Setup of package 'hos t' at packages/System/Host.yaml, line 1 failed: Invalid package definition 'hos t': invalid slug hos t (try hos_t). Package will not be initialized` during bootstrap and config reload.
3. Ignore packages that fail this validation, but loads the rest

Config reload check produces a message like this:
![grafik](https://github.com/home-assistant/core/assets/57918757/4cec234f-1bad-4707-b972-fafa454457b3)

Actual config reload adds this message to system log.

How do you feel about this?

Reply:
I updated the code and the title and description of this PR to basically reflect what I wrote in my last message.